src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs

```diff
@@ -0,0 +1,658 @@
+                        unsafe
+                        {
+                            Interop.libcurl.curl_waitfd extraFds = new Interop.libcurl.curl_waitfd {
+                                fd = _wakeupRequestedPipeFd,
+                                events = Interop.libcurl.CURL_WAIT_POLLIN,
+                                revents = 0
+                            };
+                            ThrowIfCURLMError(Interop.libcurl.curl_multi_wait(multiHandle, &extraFds, 1, FailsafeTimeoutMilliseconds, out numFds));
+                            if ((extraFds.revents & Interop.libcurl.CURL_WAIT_POLLIN) != 0)
+                            {
+                                // We woke up (at least in part) because a wake-up was requested.  Read the data out of the pipe 
+                                // to clear it. It's possible but unlikely that there will be tons of extra data in the pipe, 
+                                // more than we end up reading out here (it's unlikely because we only write a byte to the pipe when 
+                                // transitioning from 0 to 1 incoming request or when cancellation is requested, and that would
+                                // need to happen many times in a single iteration).  In that unlikely case, we'll simply loop 
+                                // around again as normal and end up waking up spuriously from the next curl_multi_wait.  For now, 
+                                // this is preferable to making additional syscalls to poll and read from the pipe).
+                                const int ClearBufferSize = 4096; // some sufficiently large size to clear the pipe in any normal case
+                                byte* clearBuf = stackalloc byte[ClearBufferSize];
+                                while (Interop.CheckIo((long)Interop.libc.read(_wakeupRequestedPipeFd, clearBuf, (IntPtr)ClearBufferSize)));
+                            }
+                        }
+
+                        // PERF NOTE: curl_multi_wait uses poll (assuming it's available), which is O(N) in terms of the number of fds 
+                        // being waited on. If this ends up being a scalability bottleneck, we can look into using the curl_multi_socket_* 
+                        // APIs, which would let us switch to using epoll by being notified when sockets file descriptors are added or 
+                        // removed and configuring the epoll context with EPOLL_CTL_ADD/DEL, which at the expense of a lot of additional
+                        // complexity would let us turn the O(N) operation into an O(1) operation.  The additional complexity would come
+                        // not only in the form of additional callbacks and managing the socket collection, but also in the form of timer
+                        // management, which is necessary when using the curl_multi_socket_* APIs and which we avoid by using just
+                        // curl_multi_wait/perform.
```

Comment:
Until we do use epoll, there's little value in the extra complexity brought by switching to ```curl_multi_socket_*```... in fact there's extra cost to it because we end up needing to make many more P/Invoke calls and the like.  I think if we're going to do the switch, it makes sense to hold off on that until a separate PR that switches from ```curl_multi_perform/wait``` to ```curl_multi_socket_*``` and ```epoll``` all together, otherwise we're paying without any gains.  And such a switch should be a relatively simple drop-in replacement with this revised design:
- At the beginning of the worker after creating the multi handle, we epoll_create an epoll instance.
- We register for the socket callback function, and when it's invoked (synchronously, so no extra locking) we use epoll_ctl with EPOLL_CTL_ADD/DEL to add or remove the socket from being tracked.
- We register the timer functions as was done previously, but when the timer fires, it simply calls RequestWakeup to make sure the processing loop goes around again.
- Instead of curl_multi_wait, we use epoll.
- Instead of curl_multi_perform, we iterate through the set fds from epoll and use curl_multi_socket_action.  Etc.
- When the worker is about to end, we delete the epoll instance.
- Etc.

It's very possible we'll want to pull back in some of the deleted code (e.g. the timer interop stuff), but we can grab that from the history, and if/when we do want to do this, we can layer it on top of a foundation that I think puts us in a more reliable place.

Let me know if you disagree.  If you think it's really important to do as part of this PR, I can look at adding an additional commit to do it.